Make orderbook depth change dynamically with layout

rjpeterson opened the pull request:
This change causes the depth value passed to the orderbook component to change in response to the height of the resizable container.
This allows users to show more or less of the depth of the orderbook depending on their preference.
Old behavior was to always show a depth of 8 rows regardless of the size of the container.
![mango-orderbook-resize](https://user-images.githubusercontent.com/4095852/124239622-7b720700-db4c-11eb-8b8c-679fea98e163.gif)

rjpeterson committed:
Make orderbook size change dynamically with layout

This change causes the depth value passed to the orderbook component to change in response to the height of the resizable container.
This allows users to show more or less of the depth of the orderbook depending on their preference.
Old behavior was to always show a depth of 8 rows regardless of the size of the container.

mschneider commented on yarn.lock:
```diff
@@ -2515,9 +2515,9 @@ boolbase@^1.0.0, boolbase@~1.0.0:
   resolved "https://registry.yarnpkg.com/boolbase/-/boolbase-1.0.0.tgz#68dff5fbe60c51eb37725ea9e3ed310dcc1e776e"
   integrity sha1-aN/1++YMUes3cl6p4+0xDcwed24=
 
-"borsh@git+https://github.com/defactojob/borsh-js.git#field-mapper":
```
can you revert changes to this file? doesn't seem to be needed

rjpeterson replied:
Changes to yarn.lock have been reverted

mschneider commented:
really cool change! do you think it would be difficult to maximize the amount of space used for the order book display a little more? sometimes there seems to be a little extra space in the bottom, that could be used for another row (especially with the side-by-side layout)

rjpeterson committed:
revert changes to yarn.lock

rjpeterson commented:
> really cool change! do you think it would be difficult to maximize the amount of space used for the order book display a little more? sometimes there seems to be a little extra space in the bottom, that could be used for another row (especially with the side-by-side layout)

Sure, I'll take another look and try to optimize it a bit more for the side-by-side layout.

I'm also aware that the top-and-bottom layout sometimes has empty space at the bottom as well, but that is due to the fact that in that layout rows get added two at a time, one in bids and one in asks, so is harder to take up all available space without having an odd number of rows in one and even in the other.

rjpeterson commented:
One other problem I have run into with this feature (that I'm unsure the best way to deal with as I'm not very familiar with react-grid-layout) is that a resizing of the browser window past a breakpoint will trigger a change in layout, but not a corresponding change in orderbook depth. I am unsure how to listen for a breakpoint change in order to trigger a depth recalculation.

rjpeterson committed:
adjust depth formula, add assymetric depth in vertical mode

The depth of the orderbook now more closely matches the size of the container.
The scrollbar has also been disabled in the orderbook panel to enable the space to be used as efficiently as possible without the annoyance of the scrollbar appearing.
The vertical orientation of the orderbook now allows an odd number of ask rows and an even number of bid rows if it is able to use more of the available space.

mschneider commented:
@tjshipe might know

mschneider commented:
i managed to repro the issue with the breakpoints, guess this is good to go, once we figure that one out

rjpeterson committed:
Make orderbook depth adjust on breakpoint change

Add orderbookDepth state which is managed by onBreakpointChange and onLayoutChange functions.

rjpeterson commented:
The orderbook depth now reacts to breakpoint changes as well as layout changes.

rjpeterson commented:
One problem I've just noticed is that since layouts are saved between sessions, the depth for each breakpoint will need to be as well.

rjpeterson committed:
Add orderbookDepth value to useMangoStore

rjpeterson committed:
Merge branch 'main' into resizable-orderbook

mschneider commented:
looks like you are missing a depth recalculation on page reload

![image](https://user-images.githubusercontent.com/144124/124401577-6ce14680-dd2a-11eb-9285-8acc4b905e32.png)

rjpeterson committed:
Simplify state management and move orderbook adjustments into useEffect

Remove orderbook state object from useMangoStore
This change solves the issue of the orderbook not adjusting on a new session, as well eliminates instances of repeated calls to adjustOrderBook()

rjpeterson committed:
Merge branch 'blockworks-foundation:main' into resizable-orderbook

rjpeterson committed:
Merge branch 'resizable-orderbook' of github.com:rjpeterson/mango-ui-v2 into resizable-orderbook

rjpeterson commented:
> looks like you are missing a depth recalculation on page reload
> 

The depth now recalculates on reload. I believe all bugs are squashed.

mschneider commented:
Looks like we still have compile errors & conflicts with the base branch, i'll have a look once those are fixed

rjpeterson committed:
Fix lodash compile errors

mschneider commented:
@rjpeterson could you have a shot at bringing this over to v3?

rjpeterson commented:
Yeah I can do that. I won't be able to start on it until mid next week
though as I'll be away from the computer till then.
